iXBRLViewerPlugin/viewer/src/js/outline.js

```diff
@@ -0,0 +1,153 @@
+                if (!(elr in longestRun) || longestRun[elr] < runLength[elr]) {
+                    longestRun[elr] = runLength[elr];
+                    longestRunStart[elr] = runStart[elr];
+                }
+                delete runLength[elr];
+                delete runStart[elr];
+            }
+        }
+    }
+
+    // End of document, check if any current runs are the longest run for the
+    // ELR.
+    for (const elr in runLength) {
+        if (!(elr in longestRun) || longestRun[elr] < runLength[elr]) {
+            longestRun[elr] = runLength[elr];
+            longestRunStart[elr] = runStart[elr];
+        }
+    }
+
+    this.sections = longestRunStart;
+}
+
+// Returns true if a fact participates in the given presentation group.
+DocumentOutline.prototype.factInGroup = function (fact, elr) {
+    // Roots are abstract so no need to check for concepts with outgoing
+    // relationships only.
+    if (this._report.getParentRelationshipsInGroup(fact.conceptName(), "pres", elr).length == 0) {
+        return false;
+    }
+    const fd = fact.dimensions();
+    const dm = this.dimensionMap[elr];
+    // Check all dimensions specified in this ELR
+    for (const [dim, spec] of Object.entries(dm)) {
+        // If a fact has a dimension, it must be in the list of permitted
+        // members, otherwise, the default member must be allowed
+        if ((dim in fd) ? !(fd[dim] in spec.members) : !spec.allowDefault) {
+            return false;
+        }
+    }
+    return true;
+}
+
+// Build a map of ELRs to dimensional information:
+//   { elr: 
+//      { dimensionQName: 
+//          { 
+//              allowDefault: true, 
+//              members: { 
+//                  memberQName: true
+//              } 
+//          } 
+//      } 
+//   } 
+DocumentOutline.prototype._buildDimensionMap = function () {
+    const groups = this._report.relationshipGroups("pres");
+    var dimensionMap = {};
+    for (const elr of groups) {
+        dimensionMap[elr] = {};
+        for (const root of this._report.relationshipGroupRoots("pres", elr)) {
+            this.buildDimensionMapFromSubTree(dimensionMap[elr], "pres", elr, null, root);
+        }
+    }
+    return dimensionMap;
+}
+
+DocumentOutline.prototype.buildDimensionMapFromSubTree = function(dimensionMap, arcrole, elr, dimension, conceptName) {
+    var children = this._report.getChildRelationships(conceptName, arcrole);
+    if (!(elr in children)) {
+        return
+    }
+    const c = this._report.getConcept(conceptName);
+    if (c.isDimension()) {
+        dimension = conceptName;
+        dimensionMap[dimension] = { members: {}, allowDefault: false};
+    }
+    for (var rel of children[elr]) {
+        if (dimension) {
+            if (this._report.dimensionDefault(dimension) == rel.t) {
+                dimensionMap[dimension].allowDefault = true;
+            }
+            else {
+                dimensionMap[dimension].members[rel.t] = true;
```

Comment:
I'm a bit lost in the XBRL-vs-JS data structures here, but I see `"pres"`.  Is this ultimately trying to build the list of dimension members based on the presentation rather than definition linkbase?

Reply:
Yes - the only "real" dimensional information used is the `dimension-default` relationship.

The problem with using the definition relationships is that you have to make the assumption that there is correspondence between the presentation and dimensional ELRs, as opposed to the assumption that we make here, which is that the presentation mirrors the dimensional structure.

Doing it based on the presentation is consistent with what the [EFM](https://www.sec.gov/info/edgar/specifications/edgarfm-vol2-v59.pdf) describes for the EDGAR renderer, e.g. (6.24.2)

> If a context segment has an axis with a member for which the _axis does not have that member as a descendant in the presentation group_ does not appear in the base set, then the context will not be selected (it will be filtered out).

Reply:
I agree the EFM requires it, but what about other filing authorities?  It seems somewhat fragile to rely on that.

Anyway, I guess this is all just heuristics, so if you/others are comfortable with it, then ok.  Perhaps just add a comment to this code that we're relying on the presentation to match the definition, and if we find some set of filings that doesn't conform to that, we could in theory make this more robust by using the definition.

Reply:
ESEF appear to mandate the same structure as SEC.

The UK taxonomies don't, but equally, there's no correspondence between presentation and definition ELRs there, and the document outline is imperfect on non-extension taxonomies anyway, as it'll list irrelevant sections that happen to include a fact that is reported in a different section.

I've added a comment to the code.